docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,354 @@
+
+let bucket = new cloud.Bucket();
+
+let client = new typescript.Client(
+  name: "@acme/infra-client",
+  lifts: @lift({ bucket: [put] }),
+);
+```
+
+In order to access the lifted objects at runtime, the TypeScript client needs to know how it should obtain the initialization data for instantiating the inflight clients (e.g. resource ARNs and other preflight state). This information is stored in client.context in preflight.
+When the client is used later (inflight), the environment variable `WING_CONTEXT` is expected to the same value previously stored in client.context.
+
+The context is just a string value (base64 encoded JSON), so you can store it in any way that is convenient for your deployment process.
+
+For example, it could be stored in a bucket:
+
+```wing
+bring cloud;
+bring clients;
+bring fs;
+
+let client = new typescript.Client(
+  name: "@acme/infra-client",
+);
+
+// The deployer of the client host will need to retrieve/link this data later
+let clientContext = new cloud.Bucket();
+clientContext.addObject("ctx", client.context);
+```
+
+When compiling to non-simulator targets, the client will store the expected permissions to be used elsewhere.
+For example, in `tf-aws`, the client will generate a policy to be attached to a role or user. The name of the policy is required to be set in the `wing.toml` file.
+
+```toml
+[tf-aws]
+client_policy = "MyClientPolicy"
+```
+
+After the client is generated and published, it can be used in any TypeScript project:
+
+```ts
+import { Client } from "@acme/infra-client";
+
+// Will use the environment variable WING_CONTEXT to resolve the context
+const client = new Client();
+
+// Optionally, you can pass the context directly to bypass the environment variable
+// const client = new Client({ context: "..." });
+
+await client.bucket.put("key", "value");
+```
+
+The client code is the same as the generated javascript as if it were used within wing itself!
+
+## Implementation Details
+
+### Client context data
+
+The client context data is a JSON object that contains all the necessary information to connect to the resources.
+It's read through the environment variable `WING_CONTEXT` and is expected to be a base64 encoded.
+
+For now, the only information that needs to be stored are environment variables.
+This corresponds to the only API available on `IInflightHost`, which is `addEnvironment`, which is how these values are added:
+
+```json
+{
+  "version": "1",
+  "env": {
+    "BUCKET_HANDLE_a107cec5": "abcxyz"
+  }
+}
+```
+
+The client itself is effectively a proxy of an unseen/unavailable `IInflightHost`, so any expressivity that interface has must be handled by the client.
+
+### Permissions for the simulator
+
+The client is effectively an `IInflightHost`, so it any resource lifted into it registers the permissions it needs.
+This is implemented on a per-target basis to expose the permissions in a useful way.
+For the `sim` target this problem is (currently) much more difficult.
```

Comment:
We have permission modeling for `sim`, no? Should work like `cloud.Function`s.

Reply:
If you're running a simulator in one process, you can't later "deploy" something else to it. Our sim permissions model only works for things deployed into it.
Tackling this would be a big lift so I lift it out for now

Reply:
After some prototyping I realized this actually works really well without much effort, I updated the RFC to reflect the reality